File: tools/convert_esri_fgdb.py
import argparse
import sys
import os
import sqlite3
import pandas as pd
import json
import arcpy

db_path = r'H:\temp\mlra27_soil.sqlite'
out_dir = r'H:\temp'
cxn = sqlite3.connect(db_path)
cxn.enable_load_extension(True)
cxn.execute("SELECT load_extension('mod_spatialite');")
# cxn.conn.execute("PRAGMA foreign_keys = ON;")

fgdb_path = r"H:\temp\temp.gdb"
if not os.path.isdir(fgdb_path):
    arcpy.management.CreateFileGDB(
        out_folder_path=os.path.dirname(fgdb_path),
        out_name=os.path.basename(fgdb_path),
        out_version="CURRENT"
    )

tabular_list = [
    'mdstatdommas',
    'mdstattabs',
    'sacatalog',
    'sdvalgorithm',
    'sdvattribute',
    'sdvfolder',
    'featdesc',
    'legend',
    'mdstatdomdet',
    'mdstatidxmas',
    'mdstatrshipmas',
    'mdstattabcols',
    'sdvfolderattribute',
    'distlegendmd',
    'distmd',
    'distinterpmd',
    'laoverlap',
    'legendtext',
    'mapunit',
    'mdstatidxdet',
    'mdstatrshipdet',
    'sainterp',
    'component',
    'muaggatt',
    'muaoverlap',
    'mucropyld',
    'mutext',
    'chorizon',
    'cocanopycover',
    'cocropyld',
    'codiagfeatures',
    'coecoclass',
    'coeplants',
    'coerosionacc',
    'coforprod',
    'cogeomordesc',
    'cohydriccriteria',
    'cointerp',
    'comonth',
    'copmgrp',
    'copwindbreak',
    'corestrictions',
    'cosurffrags',
    'cotaxfmmin',
    'cotaxmoistcl',
    'cotext',
    'cotreestomng',
    'cotxfmother',
    'chaashto',
    'chconsistence',
    'chdesgnsuffix',
    'chfrags',
    'chpores',
    'chstructgrp',
    'chtext',
    'chtexturegrp',
    'chunified',
    'coforprodo',
    'copm',
    'cosoilmoist',
    'cosoiltemp',
    'cosurfmorphgc',
    'cosurfmorphhpp',
    'cosurfmorphmr',
    'cosurfmorphss',
    'chstruct',
    'chtexture',
    'chtexturemod',
]

for tbl in tabular_list:
    sql = f'SELECT * FROM {tbl};'
    json_path = os.path.join(out_dir, f'{tbl}.json')
    csv_path = os.path.join(out_dir, f'{tbl}.csv')
    df = pd.read_sql_query(sql, cxn)
    df.spatial.to_table(os.path.join(fgdb_path, tbl))
    # df.to_json(outpath, orient='records', lines=True)
    df.to_csv(csv_path, index=False)

    recs = df.to_records(index=False)
    arcpy.da.NumPyArrayToTable(recs, os.path.join(fgdb_path, tbl))

    arcpy.conversion.JSONToFeatures(
        in_json_file=outpath,
        out_features=os.path.join(fgdb_path, tbl),
        geometry_type="POLYGON"
    )




feature_list = ['featline', 'featpoint', 'muline', 'mupoint', 'mupolygon', 'sapolygon']

File: tools/convert_esri_fgdb.R
library(sf)
library(DBI)
library(RSQLite)
library(glue)
library(tibble)

db_path <- file.path("C:/Users/wlieurance/Documents/MLRA_27_NV/Database",
                     "mlra27_soil.sqlite")
out_dir <- "C:/Users/wlieurance/Documents/MLRA_27_NV/Database"
con <- dbConnect(RSQLite::SQLite(), db_path)
dbExecute(con, "SELECT load_extension('mod_spatialite');")
# cxn.conn.execute("PRAGMA foreign_keys = ON;")

fgdb_path = "C:/Users/wlieurance/Documents/MLRA_27_NV/Database/temp.gdb"

tabular_list <- c(
  "mdstatdommas",
  "mdstattabs",
  "sacatalog",
  "sdvalgorithm",
  "sdvattribute",
  "sdvfolder",
  "featdesc",
  "legend",
  "mdstatdomdet",
  "mdstatidxmas",
  "mdstatrshipmas",
  "mdstattabcols",
  "sdvfolderattribute",
  "distlegendmd",
  "distmd",
  "distinterpmd",
  "laoverlap",
  "legendtext",
  "mapunit",
  "mdstatidxdet",
  "mdstatrshipdet",
  "sainterp",
  "component",
  "muaggatt",
  "muaoverlap",
  "mucropyld",
  "mutext",
  "chorizon",
  "cocanopycover",
  "cocropyld",
  "codiagfeatures",
  "coecoclass",
  "coeplants",
  "coerosionacc",
  "coforprod",
  "cogeomordesc",
  "cohydriccriteria",
  "cointerp",
  "comonth",
  "copmgrp",
  "copwindbreak",
  "corestrictions",
  "cosurffrags",
  "cotaxfmmin",
  "cotaxmoistcl",
  "cotext",
  "cotreestomng",
  "cotxfmother",
  "chaashto",
  "chconsistence",
  "chdesgnsuffix",
  "chfrags",
  "chpores",
  "chstructgrp",
  "chtext",
  "chtexturegrp",
  "chunified",
  "coforprodo",
  "copm",
  "cosoilmoist",
  "cosoiltemp",
  "cosurfmorphgc",
  "cosurfmorphhpp",
  "cosurfmorphmr",
  "cosurfmorphss",
  "chstruct",
  "chtexture",
  "chtexturemod",
  "featline", 
  "featpoint", 
  "muline", 
  "mupoint", 
  "mupolygon", 
  "sapolygon",
  "ecopolygon",
  "ecogrouppolygon",
  "ecogroup",
  "ecogroup_meta",
  "coecoclass_area",
  "coecoclass_codominant",
  "coecoclass_mapunit_ranked",
  "coecoclass_mudominant",
  "coecoclass_plantprod",
  "coecoclass_unique",
  "coecoclass_wide",
  "coeplants_gh",
  "component_dominant",
  "component_mapunit_ranked",
  "component_plantprod",
  "component_unique",
  "component_wide",
  "ecogroup_area",
  "ecogroup_detail",
  "ecogroup_mapunit_ranked",
  "ecogroup_mudominant",
  # "ecogroup_plantprod",
  "ecogroup_unique",
  "ecogroup_wide"
)

tbls_sql <- "SELECT name FROM sqlite_master WHERE type IN ('table', 'view');"
sqlite_tbls <- dbGetQuery(con, tbls_sql)

start <- 79
for (i in start:length(tabular_list)) {
  tbl <- tabular_list[i]
  if (tbl %in% sqlite_tbls$name) {
    print(paste0("Importing ", tbl, "..."))
    # sql <- glue("SELECT * FROM {tbl};")
    df <- st_read(dsn = db_path, layer = tbl, as_tibble = TRUE, 
                  drivers = "SQLite", quiet = TRUE)
    if (nrow(df) > 0) {
      st_write(obj = df, dsn = fgdb_path, layer = tbl, driver = "OpenFileGDB", 
               append = FALSE, quiet = TRUE)
    } else {
      print(paste(tbl, "empty. Skipping..."))
    }
  } else {
    print(paste(tbl, "not in database. Skipping..."))
  }
}
